flink-runtime/src/test/java/org/apache/flink/runtime/blob/PermanentBlobCacheSizeLimitTest.java

```diff
@@ -0,0 +1,251 @@
+            for (int i = 0; i < 3; i++) {
+                key[i] = put(server, jobId[i], data[i], blobType);
+                assertNotNull(key[i]);
+
+                readFileAndVerifyContent(cache, jobId[i], key[i], data[i]);
+                file[i] = getFile(cache, jobId[i], key[i]);
+                assertTrue(file[i].exists());
+            }
+
+            // Since the size limit of the cache is the size of 2 blobs,
+            // the first blob is removed and the second one remains
+            assertFalse(file[0].exists());
+            assertTrue(file[1].exists());
+
+            // Get the second blob once again, make the third blob to be the last recently used
+            readFileAndVerifyContent(cache, jobId[1], key[1], data[1]);
+
+            // Then get the fourth blob, make sure the third blob is replaced.
+            // Here we cannot get the first blob again, because for transient blob,
+            // BlobServer will delete the blob once it's transmitted.
+            key[3] = put(server, jobId[3], data[3], blobType);
+            readFileAndVerifyContent(cache, jobId[3], key[3], data[3]);
+            file[3] = getFile(cache, jobId[3], key[3]);
+
+            assertTrue(file[1].exists());
+            assertTrue(file[3].exists());
+            assertFalse(file[2].exists());
+        }
+    }
+
+    @Test
+    public void testTrackSizeLimitAndDeleteExcessConcurrently() throws Exception {
+
+        final BlobKey.BlobType blobType = BlobKey.BlobType.PERMANENT_BLOB;
+
+        final Configuration config = new Configuration();
+        config.setString(
+                BlobServerOptions.STORAGE_DIRECTORY, temporaryFolder.newFolder().getAbsolutePath());
+
+        final int concurrency = 3;
+        final ExecutorService executor = Executors.newFixedThreadPool(concurrency);
+
+        // The size limit is the size of 2 blobs
+        final int sizeLimit = 2;
+
+        try (BlobServer server = new BlobServer(config, new VoidBlobStore());
+                BlobCacheService cache =
+                        initBlobCacheServiceWithSizeLimit(
+                                config,
+                                new VoidBlobStore(),
+                                new InetSocketAddress("localhost", server.getPort()),
+                                sizeLimit * BLOB_SIZE)) {
+
+            server.start();
+
+            // Initialize the data structures
+            JobID[] jobId = new JobID[concurrency];
+            BlobKey[] key = new BlobKey[concurrency];
+            File[] file = new File[concurrency];
+
+            byte[] blobData = new byte[BLOB_SIZE];
+            RANDOM.nextBytes(blobData);
+
+            // Put all the BLOBs into the blob server one by one
+            for (int i = 0; i < concurrency; i++) {
+                jobId[i] = new JobID();
+
+                key[i] = put(server, jobId[i], blobData, blobType);
+                assertNotNull(key[i]);
```

Comment:
all `blobData` are the same. Why are we not doing the same as in `testTrackSizeLimitAndDeleteExcess()`?